### ISSUE TITLE
[apex] Setting the Authorization header is reported as hard coded credentials as per "Avoid Using Hard Coded Credentials for Authentication" rule

### ISSUE THREAD

**Issue by Syed-Salesforce** [pmd/pmd#4773]
**Affects PMD Version:** 

**Rule:**Avoid Using Hard Coded Credentials for Authentication

https://docs.pmd-code.org/latest/pmd_rules_apex_security.html#apexsuggestusingnamedcred

**Description:**

Setting the Authorization header to the bearer access token like below is always getting flagged as hard coding, when there’s nothing hard coded about it.

The code is setting an “Authorization” header, passing an access token that was obtained earlier in the code.

Can you please confirm if this rule is highlighting the syntax 'Bearer'?

**Code Sample demonstrating the issue:**

 
```java
req.setHeader('Authorization', 'Bearer ' + tokenResponse);
```

**Expected outcome:** 

PMD reports a violation at line ..., but that's wrong. That's a false positive.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]* : Ant

**Comment by Syed-Salesforce** [comment]
Hi Team,

Can someone check and confirm on this request?

Thank you

**Comment by adangel** [comment]
Hi @Syed-Salesforce ,
this rule is about preferring Named Credentials over manually coding any authorization. In that sense, no matter where your "bearer" value is coming from, this rule should flag any usage of "request.setHeader("Authorization", ...)". So, I don't think, this is a false positive and it is correct.

The name "bearer" sounds like you use OAuth or something similar for auths. Quoting https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_callouts_named_credentials.htm:

> Coding OAuth is even more complex and is an ideal use case for named credentials.

I'd suggest to configure Named Credentials and use that via `req.setEndpoint('callout:My_Named_Credential/some_path')`.
